### PR TITLE
[SYCL] Emit a warning on constexprs captured in SYCL kernels

### PR DESCRIPTION
**Pull request by srividya-sundaram**
In some cases Clang FE captures `constexprs`  variables into a lambda. This can cause performance problems in case of SYCL, because

- All captured variables become a kernel parameter, which shouldn't be the case for `constexprs`
- Captured `constexpr` will not appear as a `compile-time` constant to the compiler 

This patch warns users that a `constexpr` has been captured as a kernel argument and is not a `compile-time` constant.
The exact wordings of the warning message is open for wordsmithing.

### CODE REVIEW

**Comment by srividya-sundaram** [comment]
clang format fails for `warning-flags.c`, but formatting it results in modifying the file in a weird way by appending the RUN lines together.  Ex: see [here](https://github.com/intel/llvm/blob/2f4908ee5fbb7a0ac8cad9bfbc40f11e01f0352b/clang/test/Misc/warning-flags.c)

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (premanandrao) [inline clang/include/clang/Basic/DiagnosticGroups.td:1258]: Were you planning to add more diagnostics to this group?  If not, I think we could defer having a group for this until later.

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: I think it will be helpful to list the captured variable name in the diagnostic.

File: clang/test/Misc/warning-flags.c
Comment (premanandrao) [inline clang/test/Misc/warning-flags.c:21]: This one's interesting.  The last time I had to do this, I instead found a flag to tie the diagnostic under.   But I can't think of an existing one for this diagnostic; nor do I think it is worth adding a flag for just this.  Advice, @Fznamznon?

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (srividya-sundaram) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I didn't have a new group added initially but then "warning-flags.c" failed and looking at that file's description, it suggested that I add anew group for warnings without a dedicated flag.
See: [L1516](https://github.com/intel/llvm/blob/sycl/clang/test/Misc/warning-flags.c#L15)

File: clang/test/Misc/warning-flags.c
Comment (srividya-sundaram) [reply clang/test/Misc/warning-flags.c:21]: Right, I didn't think it warranted a separate flag for this case, so I deferred from adding one.
But then for newly added warnings without a flag, it required me to add a warning group.
None of the existing warning groups seemed to work for this case, so I ended up adding a new one.
If I can re-use any existing warning groups, please let me know. 

`If you add a new warning without a flag, this test will fail.  To fix this test, simply add a warning group to that warning.`

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I can't think of an existing flag to reuse at the moment, so this is okay.  But it would be cleaner to avoid it.

File: clang/test/Misc/warning-flags.c
Comment (srividya-sundaram) [reply clang/test/Misc/warning-flags.c]: **Update**: After running check-all tests locally and based on the temp file generated (build/tools/clang/test/Misc/Output/warning-flags.c.tmp), I reverted the changes made in this test.
The test passes and fixed the clang-format error as well.

File: clang/lib/Sema/SemaSYCL.cpp
Comment (Fznamznon) [inline clang/lib/Sema/SemaSYCL.cpp:3053]: I don't see a point in calling `Capture.isImplicit()` two times.
```suggestion
      if (Capture.isImplicit()) {
        if (Capture.capturesThis()) {
          Diag(Capture.getLocation(), diag::err_implicit_this_capture);
          Diag(CallLoc.getBegin(), diag::note_used_here);
          KernelFunc->setInvalidDecl();
        }
        if (Capture.capturesVariable() {
          if (Capture.getCapturedVar()->isConstexpr())
            Diag(Capture.getLocation(), diag::warn_constexpr_kernel_arg)
              << Capture.getCapturedVar()->getNameAsString();
        }
      }
```

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I feel adding this group adds at least -W flag for the new diagnostic. "warning-flags.c" test says that we shouldn't add the new warnings without flags, i.e. we probably need some group anyway. So, I think we need to understand what do we want to do. For example we have some -sycl-strict diagnostic group, but warnings from this group are disabled by default by the Driver, if we are fine with this behavior, we can re-use sycl-strict group. If we want our warning to be enabled by default, we could add a special group for it. Or, for example, this warning says about possible non-obvious performance degradation to user, I think we may have another cases of similar non-obvious performance degradations as well as another warnings for such cases, so we can allocate the group for them.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:7453]: Could you please keep lines short?
```suggestion
    "constexpr '%0' is captured and is not a compile-time constant">,
    InGroup<ConstExprKernelArg>;
```

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (Fznamznon) [inline clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:1]: Is `-Wsycl-strict` flag necessary here?
```suggestion
// RUN: %clang_cc1 -fsycl -fsycl-is-device -internal-isystem %S/Inputs -fsyntax-only -sycl-std=2020 -verify %s
```

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (elizabethandrews) [inline clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: Do you need 'Id' for this test? Can the test be made simpler?

Comment (Fznamznon) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: The point is to make constexpr odr-used. In this test it is odr-used because operator* takes a reference. Any other test which makes constexpt odr-used will also work here.

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I feel this warning should be enabled by default, as like you say, it detects something immediately non-obvious and likely helpful for performance reasons. Adding a group for this purpose would be great.   Does the name have to be decided now?  Will choosing the group name at a later stage be considered "breaking" as the old spelling will no longer work?

Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: > Does the name have to be decided now? Will choosing the group name at a later stage be considered "breaking" as the old spelling will no longer work?

I think it is better to choose the name now, so we don't break user's code in the future.

Comment (srividya-sundaram) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: Any suggestions for the generic group name that would capture all non-obvious warnings? <nonobvious-perf-degradation>?

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (srividya-sundaram) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:1]: It is not required for this test. I will remove it.

**Comment by srividya-sundaram** [comment]
@bader The test failures reported are all related to sycl-post-link tool in `/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/llvm/test/tools/sycl-post-link/`

I verified the tests locally and they all pass. Could you clarify if this a buildbot related issue?
Also, clang-format check has been queued for a long time now.

**Comment by Fznamznon** [comment]
> @bader The test failures reported are all related to sycl-post-link tool in `/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/llvm/test/tools/sycl-post-link/`
> 

@AlexeySachkov , could you please take a look? Can it be caused by https://github.com/intel/llvm/commit/e481174da6470e055e4b401e86949ed801bb6c16 ?

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I'm not really good at this but, how about something like `perf-impact`?

**Comment by AlexeySachkov** [comment]
> > @bader The test failures reported are all related to sycl-post-link tool in `/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/llvm/test/tools/sycl-post-link/`
> 
> @AlexeySachkov , could you please take a look? Can it be caused by [e481174](https://github.com/intel/llvm/commit/e481174da6470e055e4b401e86949ed801bb6c16) ?

Sure, I will take a look. It seems like pulldown has landed before my PR got merged, but after I had launched pre-commit: there was a change in typed attributes for function arguments, I just need to re-generate LLVM IR for those tests

**Comment by AlexeySachkov** [comment]
> > > @bader The test failures reported are all related to sycl-post-link tool in `/localdisk2/sycl_ci/buildbot/worker/sycl-ubu-x64-pr/llvm.src/llvm/test/tools/sycl-post-link/`
> > 
> > 
> > @AlexeySachkov , could you please take a look? Can it be caused by [e481174](https://github.com/intel/llvm/commit/e481174da6470e055e4b401e86949ed801bb6c16) ?
> 
> Sure, I will take a look. It seems like pulldown has landed before my PR got merged, but after I had launched pre-commit: there was a change in typed attributes for function arguments, I just need to re-generate LLVM IR for those tests

Should be fixed by #2846

File: clang/include/clang/Basic/DiagnosticGroups.td
Comment (srividya-sundaram) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: Works for me, unless @premanandrao or @elizabethandrews have other suggestions. If I don't hear any other options by tomorrow, I will update the group name to `perf-impact`.
Thanks, @Fznamznon

Comment (premanandrao) [reply clang/include/clang/Basic/DiagnosticGroups.td:1258]: I am okay with that spelling.

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (Fznamznon) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: Was it resolved?

Comment (srividya-sundaram) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: This was marked resolved cause the example makes ODR-use of `constexpr` var 'V' which makes lambda to capture it. So the struct Id wasn't added without a reason. But I can try to make the test case simpler and update.

Comment (Fznamznon) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: > But I can try to make the test case simpler and update.

If I understand correctly, @elizabethandrews were asking for it.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (Fznamznon) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:7453]: How about something like this:
```suggestion
    "captured constexpr '%0' will appear as regular kernel argument in device code">,
```
It sounds a bit more clear

Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:7453]: I think we can omit 'regular' and just say something like - 

 "captured constexpr '%0' will be a kernel argument in device code"

I'm not too particular about this and so feel free to choose what you prefer.

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (elizabethandrews) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:34]: I didn't understand the point of this test initially. After @Fznamznon's explanation, I am okay with this if none of the other reviewers have an issue . However, wouldn't a simple pointer assignment in kernel also achieve required test?
`const unsigned *ptr = &V;`

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:7453]: I'm fine with both variants.

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (srividya-sundaram) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp]: `const unsigned *ptr = &V;` is ODR-used as well. Updated a simplified version of the test.

**Review by Fznamznon — APPROVED** [review]

**Review by elizabethandrews — APPROVED** [review]
Thanks!

**Comment by srividya-sundaram** [comment]
@bader This is ready to be merged.

**Comment by bader** [comment]
It sounds like the compiler drops `constexpr` semantics. Could you clarify why this is not a compiler bug, please?

**Comment by srividya-sundaram** [comment]
> It sounds like the compiler drops constexpr semantics. Could you clarify why this is not a compiler bug, please?

Compiler captures a `constexpr` as a kernel argument only when `constexpr` is ODR-used inside the lambda.
In other cases, it is treated as a compile-time constant as it should be.
This is why the compiler emits a warning instead of treating it as an error. 
Hope I understood your question correctly.

**Comment by premanandrao** [comment]
> It sounds like the compiler drops `constexpr` semantics. Could you clarify why this is not a compiler bug, please?

The warning generated is only for implicit captures.  Since the address of the constexpr variable is taken, it gets turned into a kernel parameter.   Otherwise, it remains constexpr.  @erichkeane says that this could potentially be addressed in the library with overloads (and that it would be difficult to do so), but I didn't quite follow that thought.

**Review by premanandrao — APPROVED** [review]

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (bader) [inline clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:23]: I don't quite get the purpose of this warning.

I think in order to understand this warning message, user must know DPC++ implementation details about [lowering lambda object into kernel function](https://intel.github.io/llvm-docs/CompilerAndRuntimeDesign.html#lowering-of-lambda-function-objects-and-named-function-objects) and how kernel function parameters are composed. Do you think it's clear for DPC++ users?

Another point it's generic C++ behavior and not SYCL specific issue:

```c++
template <typename T>
void foo(T t) {t();}
// ...
foo([=]() { const unsigned *ptr = &OdrUsedVar; }); // OdrUsedVar is captured and might cause performance issues
foo([=]() { unsigned x = 10 + OdrUsedVar; }); // OdrUsedVar is not captured. No performance issues?
```
Should we emit similar warning for any lambda expression?

It's probably more useful to implement such diagnostics in performance profiler because passing additional argument might be not a problem at all.

@premanandrao, @srividya-sundaram, @erichkeane, what do you think?

Comment (srividya-sundaram) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:23]: tagging @kbobrovs to give his feedback since this was originally brought up by him.

Comment (bader) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:23]: One more thing: this diagnostics warns that current implementation is not optimal.
I assume if compiler detects odr-use of `constexpr` it can allocate a global constant/variable in device code and kernel function object can capture global object and avoid passing it via kernel argument. This seems to be valid implementation. Right?

Comment (kbobrovs) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:23]: > It's probably more useful to implement such diagnostics in performance profiler because passing additional argument might be not a problem at all.

I suspect many people (like myself) writing code similar to
```
  constexpr int V = 137;

  auto f = [=](Id i) -> int {
    return (i * V).x;
  };
```
will expect V to be compile-time constant inside the lambda. In some cases this expectation is not true and V is captured and hence is not a compile-time constant. This can potentially lead to performance problems.

I'm not sure what 'performance compiler' is. If it can be built from this github repo - I'm OK with making this feature available only in 'performance compiler'.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td
Comment (kbobrovs) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:7453]:
```suggestion
    "captured constexpr '%0' will *not* be a compile-time constant in the device code">,
```

File: clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp
Comment (erichkeane) [reply clang/test/SemaSYCL/warn-constexpr-kernel-arg.cpp:23]: FWIW, that is a fairly well known property of constexpr: It isn't necessarily a compile time constant unless you make it so in some way.  I'm generally not in favor of a warning that tells you that the language is still working the way it is designed, particularly since there isn't a code-way to suppress this warning.

If we had some sort of perf-analysis tool, that would be a good place for that.  Have we considered seeing if the Static Analysis effort would be a better place for this?

**Comment by erichkeane** [comment]
I'm not in favor of putting a warning like this into the compiler.  Warnings are supposed to be actionable, not informative.  If there isn't a code-based way to suppress a warning (and get the behavior you want), they generally aren't put into the compiler.

The issue here is closer to a bug (though not quite?) in the library.  We're implementing conversion operators that aren't constexpr, so they cause us to need to use the variable in an ODR way, and thus need to be captured.  This is a 'feature' of the language.

IMO, if we want to warn on this, I'd suggest this go into the static analyzer.

**Comment by kbobrovs** [comment]
> Warnings are supposed to be actionable, not informative.

Agree. Possible action is to replace with `#define`. Or try to figure out the construct causing this (the `*` operator) and restructure the code (less attractive to a developer).

> This is a 'feature' of the language.

I see. But I wonder what share of SYCL developers know C++ well enough to have the right expectation of the result of the code here :) I think such warning would save many people some time and annoy just few experts. But I'll let you FE folks decide what's best here.

**Comment by srividya-sundaram** [comment]
Closing this PR since the above discussion suggests that this change should not be implemented in CFE.

**Pull request closed without merging**
